File: python/Graph_RAG-Langgraph.py
"""
Graph-RAG Implementation sử dụng LangGraph

Đây là một hệ thống RAG (Retrieval Augmented Generation) được cấu trúc dưới dạng đồ thị, bao gồm:

1. Retrieval: Truy xuất tài liệu liên quan từ vector database
2. Grading: Đánh giá độ liên quan của tài liệu
3. Rewriting: Viết lại câu hỏi nếu cần
4. Generation: Tạo câu trả lời dựa trên tài liệu

Luồng xử lý:
- Người dùng đặt câu hỏi
- Agent quyết định có cần truy xuất tài liệu không
- Hệ thống đánh giá tài liệu và quyết định generate hoặc rewrite
- Cuối cùng tạo ra câu trả lời phù hợp
"""

from langchain_community.document_loaders import WebBaseLoader
from langchain_community.vectorstores import Chroma
from langchain_openai import OpenAIEmbeddings
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_community.document_loaders import PyPDFLoader
from typing import Annotated, Sequence
from typing_extensions import TypedDict
from langchain_core.messages import BaseMessage
from langgraph.graph.message import add_messages
from typing import Annotated, Literal, Sequence
from typing_extensions import TypedDict
from langchain import hub
from langchain_core.messages import BaseMessage, HumanMessage
from langchain_core.output_parsers import StrOutputParser
from langchain_core.prompts import PromptTemplate
from langchain_openai import ChatOpenAI
from pydantic import BaseModel, Field
from langgraph.prebuilt import tools_condition
from langgraph.graph import END, StateGraph, START
from langgraph.prebuilt import ToolNode
from IPython.display import Image, display
import pprint
import os
from dotenv import load_dotenv

load_dotenv(dotenv_path="../shared/.env")

OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
if not OPENAI_API_KEY:
    raise ValueError("OPENAI_API_KEY not found in environment variables")

loader = PyPDFLoader(
    "../documents/2404.16130v1_Graph-RAG.pdf",
)
docs = loader.load()
docs[0]
text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=200)
splits = text_splitter.split_documents(docs)

# Add to vectorDB
vectorstore = Chroma.from_documents(
    documents=splits,
    collection_name="rag-chroma",
    embedding=OpenAIEmbeddings(),
)
retriever = vectorstore.as_retriever()

from langchain.tools.retriever import create_retriever_tool

retriever_tool = create_retriever_tool(
    retriever,
    "retrieve_graph_rag",
    "Search and return information about Graph-RAG.",
)

tools = [retriever_tool]

class AgentState(TypedDict):
    """Lưu trữ các tin nhắn trong quá trình tương tác"""
    messages: Annotated[Sequence[BaseMessage], add_messages]

def grade_documents(state) -> Literal["generate", "rewrite"]:
    """
    Đánh giá mức độ liên quan của tài liệu được truy xuất với câu hỏi của người dùng.
    
    Args:
        state: Trạng thái hiện tại chứa lịch sử tin nhắn
        
    Returns:
        "generate": Nếu tài liệu liên quan đến câu hỏi
        "rewrite": Nếu tài liệu không liên quan và cần viết lại câu hỏi
    """
    print("---CHECK RELEVANCE---")

    # Mô hình dữ liệu
    class grade(BaseModel):
        """Binary score for relevance check."""
        binary_score: str = Field(description="Relevance score 'yes' or 'no'")

    # LLM
    model = ChatOpenAI(temperature=0, model="gpt-4-0125-preview", streaming=True)

    # LLM với công cụ và xác thực
    llm_with_tool = model.with_structured_output(grade)

    # Prompt
    prompt = PromptTemplate(
        template="""You are a grader assessing relevance of a retrieved document to a user question. \n 
        Here is the retrieved document: \n\n {context} \n\n
        Here is the user question: {question} \n
        If the document contains keyword(s) or semantic meaning related to the user question, grade it as relevant. \n
        Give a binary score 'yes' or 'no' score to indicate whether the document is relevant to the question.""",
        input_variables=["context", "question"],
    )

    # Chain
    chain = prompt | llm_with_tool

    messages = state["messages"]
    last_message = messages[-1]

    question = messages[0].content
    docs = last_message.content

    scored_result = chain.invoke({"question": question, "context": docs})

    score = scored_result.binary_score

    if score == "yes":
        print("---DECISION: DOCS RELEVANT---")
        return "generate"

    else:
        print("---DECISION: DOCS NOT RELEVANT---")
        print(score)
        return "rewrite"

def agent(state):
    """
    Agent chính xử lý việc tương tác với người dùng và quyết định các bước tiếp theo.
    
    Args:
        state: Trạng thái hiện tại chứa lịch sử tin nhắn
        
    Returns:
        dict: Trạng thái mới với phản hồi của agent
    """
    print("---CALL AGENT---")
    messages = state["messages"]
    model = ChatOpenAI(temperature=0, streaming=True, model="gpt-4o")
    model = model.bind_tools(tools)
    response = model.invoke(messages)
    # We return a list, because this will get added to the existing list
    return {"messages": [response]}

def rewrite(state):
    """
    Viết lại câu hỏi để tạo ra một câu hỏi tốt hơn khi tài liệu không liên quan.
    
    Args:
        state: Trạng thái hiện tại chứa lịch sử tin nhắn và câu hỏi gốc
        
    Returns:
        dict: Trạng thái mới với câu hỏi đã được viết lại
    """
    print("---TRANSFORM QUERY---")
    messages = state["messages"]
    question = messages[0].content

    msg = [
        HumanMessage(
            content=f""" \n 
    Look at the input and try to reason about the underlying semantic intent / meaning. \n 
    Here is the initial question:
    \n ------- \n
    {question} 
    \n ------- \n
    Formulate an improved question: """,
        )
    ]

    # Grader
    model = ChatOpenAI(temperature=0, model="gpt-4o", streaming=True)
    response = model.invoke(msg)
    return {"messages": [response]}

def generate(state):
    """
    Tạo câu trả lời dựa trên tài liệu liên quan đã được tìm thấy.
    
    Args:
        state: Trạng thái hiện tại chứa lịch sử tin nhắn và tài liệu
        
    Returns:
        dict: Trạng thái mới với câu trả lời được tạo ra
    """
    print("---GENERATE---")
    messages = state["messages"]
    question = messages[0].content
    last_message = messages[-1]

    docs = last_message.content

    # Prompt
    prompt = hub.pull("rlm/rag-prompt")

    # LLM
    llm = ChatOpenAI(model_name="gpt-4o", temperature=0, streaming=True)

    # Post-processing
    def format_docs(docs):
        return "\n\n".join(doc.page_content for doc in docs)

    # Chain
    rag_chain = prompt | llm | StrOutputParser()

    # Run
    response = rag_chain.invoke({"context": docs, "question": question})
    return {"messages": [response]}

print("*" * 20 + "Prompt[rlm/rag-prompt]" + "*" * 20)
prompt = hub.pull("rlm/rag-prompt").pretty_print()  # Show what the prompt looks like

# Define a new graph
workflow = StateGraph(AgentState)

# Define the nodes we will cycle between
workflow.add_node("agent", agent)  # agent
retrieve = ToolNode([retriever_tool])
workflow.add_node("retrieve", retrieve)  # retrieval
workflow.add_node("rewrite", rewrite)  # Re-writing the question
workflow.add_node(
    "generate", generate
)  # Generating a response after we know the documents are relevant
# Call agent node to decide to retrieve or not
workflow.add_edge(START, "agent")

# Decide whether to retrieve
workflow.add_conditional_edges(
    "agent",
    # Assess agent decision
    tools_condition,
    {
        # Translate the condition outputs to nodes in our graph
        "tools": "retrieve",
        END: END,
    },
)

# Edges taken after the `action` node is called.
workflow.add_conditional_edges(
    "retrieve",
    # Assess agent decision
    grade_documents,
)
workflow.add_edge("generate", END)
workflow.add_edge("rewrite", "agent")

# Compile
graph = workflow.compile()

try:
    display(Image(graph.get_graph(xray=True).draw_mermaid_png()))
except Exception:
    # This requires some extra dependencies and is optional
    pass

inputs = {
    "messages": [
        ("user", "What is Graph-RAG?"),
    ]
}
for output in graph.stream(inputs):
    for key, value in output.items():
        pprint.pprint(f"Output from node '{key}':")
        pprint.pprint("---")
        pprint.pprint(value, indent=2, width=80, depth=None)
    pprint.pprint("\n---\n")

File: typescript/Graph_RAG-Langgraph.ts
import { CheerioWebBaseLoader } from "@langchain/community/document_loaders/web/cheerio";
import { RecursiveCharacterTextSplitter } from "@langchain/textsplitters";
import { MemoryVectorStore } from "langchain/vectorstores/memory";
import { OpenAIEmbeddings } from "@langchain/openai";
import { Annotation } from "@langchain/langgraph";
import { createRetrieverTool } from "langchain/tools/retriever";
import { ToolNode } from "@langchain/langgraph/prebuilt";
import { END } from "@langchain/langgraph";
import { pull } from "langchain/hub";
import { z } from "zod";
import { ChatPromptTemplate } from "@langchain/core/prompts";
import { ChatOpenAI } from "@langchain/openai";
import { AIMessage, BaseMessage } from "@langchain/core/messages";
import { StateGraph } from "@langchain/langgraph";
import { START } from "@langchain/langgraph";
import { HumanMessage } from "@langchain/core/messages";
import { PDFLoader } from "@langchain/community/document_loaders/fs/pdf";

import * as dotenv from "dotenv";

const GraphState = Annotation.Root({
  messages: Annotation<BaseMessage[]>({
    reducer: (x, y) => x.concat(y),
    default: () => [],
  }),
});

async function loadAndProcessDocuments(pdfPath: string) {
  const loader = new PDFLoader(pdfPath);
  const docs = await loader.load();

  const textSplitter = new RecursiveCharacterTextSplitter({
    chunkSize: 500,
    chunkOverlap: 50,
  });
  const docSplits = await textSplitter.splitDocuments(docs);

  // Add to vectorDB
  const vectorStore = await MemoryVectorStore.fromDocuments(
    docSplits,
    new OpenAIEmbeddings()
  );

  return vectorStore.asRetriever();
}

function shouldRetrieve(state: typeof GraphState.State): string {
  const { messages } = state;
  console.log("---DECIDE TO RETRIEVE---");
  const lastMessage = messages[messages.length - 1];

  if (
    "tool_calls" in lastMessage &&
    Array.isArray(lastMessage.tool_calls) &&
    lastMessage.tool_calls.length
  ) {
    console.log("---DECISION: RETRIEVE---");
    return "retrieve";
  }
  // If there are no tool calls then we finish.
  return END;
}

/**
 * Xác định xem Agent có nên tiếp tục dựa trên mức độ liên quan của các tài liệu được truy xuất hay không.
 * Hàm này kiểm tra xem tin nhắn cuối cùng trong cuộc hội thoại có phải là FunctionMessage hay không,
 * cho biết việc truy xuất tài liệu đã được thực hiện. Sau đó nó đánh giá mức độ liên quan của các tài liệu này
 * với câu hỏi ban đầu của người dùng bằng cách sử dụng một mô hình và bộ phân tích đầu ra được xác định trước.
 * Nếu tài liệu có liên quan, cuộc hội thoại được coi là hoàn tất. Ngược lại, quá trình truy xuất sẽ được tiếp tục.
 * @param {typeof GraphState.State} state - Trạng thái hiện tại của agent, bao gồm tất cả các tin nhắn.
 * @returns {Promise<Partial<typeof GraphState.State>>} - Trạng thái đã cập nhật với tin nhắn mới được thêm vào danh sách tin nhắn.
 */
async function gradeDocuments(
  state: typeof GraphState.State
): Promise<Partial<typeof GraphState.State>> {
  console.log("---GET RELEVANCE---");

  const { messages } = state;
  const tool = {
    name: "give_relevance_score",
    description: "Give a relevance score to the retrieved documents.",
    schema: z.object({
      binaryScore: z.string().describe("Relevance score 'yes' or 'no'"),
    }),
  };

  const prompt = ChatPromptTemplate.fromTemplate(
    `You are a grader assessing relevance of retrieved docs to a user question.
  Here are the retrieved docs:
  \n ------- \n
  {context} 
  \n ------- \n
  Here is the user question: {question}
  If the content of the docs are relevant to the users question, score them as relevant.
  Give a binary score 'yes' or 'no' score to indicate whether the docs are relevant to the question.
  Yes: The docs are relevant to the question.
  No: The docs are not relevant to the question.`
  );

  const model = new ChatOpenAI({
    model: "gpt-4o",
    temperature: 0,
  }).bindTools([tool], {
    tool_choice: tool.name,
  });

  const chain = prompt.pipe(model);

  const lastMessage = messages[messages.length - 1];

  const score = await chain.invoke({
    question: messages[0].content as string,
    context: lastMessage.content as string,
  });

  return {
    messages: [score],
  };
}

/**
 * Kiểm tra mức độ liên quan của lệnh gọi công cụ LLM trước đó.
 *
 * @param {typeof GraphState.State} state - Trạng thái hiện tại của agent, bao gồm tất cả các tin nhắn.
 * @returns {string} - Một chỉ thị "yes" hoặc "no" dựa trên mức độ liên quan của tài liệu.
 */
function checkRelevance(state: typeof GraphState.State): string {
  console.log("---CHECK RELEVANCE---");

  const { messages } = state;
  const lastMessage = messages[messages.length - 1];
  if (!("tool_calls" in lastMessage)) {
    throw new Error(
      "The 'checkRelevance' node requires the most recent message to contain tool calls."
    );
  }
  const toolCalls = (lastMessage as AIMessage).tool_calls;
  if (!toolCalls || !toolCalls.length) {
    throw new Error("Last message was not a function message");
  }

  if (toolCalls[0].args.binaryScore === "yes") {
    console.log("---DECISION: DOCS RELEVANT---");
    return "yes";
  }
  console.log("---DECISION: DOCS NOT RELEVANT---");
  return "no";
}

// Nodes

/**
 * Gọi mô hình agent để tạo phản hồi dựa trên trạng thái hiện tại.
 * Hàm này gọi mô hình agent để tạo phản hồi cho trạng thái hội thoại hiện tại.
 * Phản hồi được thêm vào tin nhắn của trạng thái.
 * @param {typeof GraphState.State} state - Trạng thái hiện tại của agent, bao gồm tất cả các tin nhắn.
 * @returns {Promise<Partial<typeof GraphState.State>>} - Trạng thái đã cập nhật với tin nhắn mới được thêm vào danh sách tin nhắn.
 */
async function agent(
  state: typeof GraphState.State,
  tools: any[]
): Promise<Partial<typeof GraphState.State>> {
  console.log("---CALL AGENT---");

  const { messages } = state;
  // Find the AIMessage which contains the `give_relevance_score` tool call,
  // and remove it if it exists. This is because the agent does not need to know
  // the relevance score.
  const filteredMessages = messages.filter((message) => {
    if (
      "tool_calls" in message &&
      Array.isArray(message.tool_calls) &&
      message.tool_calls.length > 0
    ) {
      return message.tool_calls[0].name !== "give_relevance_score";
    }
    return true;
  });

  const model = new ChatOpenAI({
    model: "gpt-4o",
    temperature: 0,
    streaming: true,
  }).bindTools(tools);

  const response = await model.invoke(filteredMessages);
  return {
    messages: [response],
  };
}

/**
 * Chuyển đổi câu truy vấn để tạo ra một câu hỏi tốt hơn.
 * @param {typeof GraphState.State} state - Trạng thái hiện tại của agent, bao gồm tất cả các tin nhắn.
 * @returns {Promise<Partial<typeof GraphState.State>>} - Trạng thái đã cập nhật với tin nhắn mới được thêm vào danh sách tin nhắn.
 */
async function rewrite(
  state: typeof GraphState.State
): Promise<Partial<typeof GraphState.State>> {
  console.log("---TRANSFORM QUERY---");

  const { messages } = state;
  const question = messages[0].content as string;
  const prompt = ChatPromptTemplate.fromTemplate(
    `Look at the input and try to reason about the underlying semantic intent / meaning. \n 
Here is the initial question:
\n ------- \n
{question} 
\n ------- \n
Formulate an improved question:`
  );

  // Grader
  const model = new ChatOpenAI({
    model: "gpt-4o",
    temperature: 0,
    streaming: true,
  });
  const response = await prompt.pipe(model).invoke({ question });
  return {
    messages: [response],
  };
}

/**
 * Tạo câu trả lời
 * @param {typeof GraphState.State} state - Trạng thái hiện tại của agent, bao gồm tất cả các tin nhắn.
 * @returns {Promise<Partial<typeof GraphState.State>>} - Trạng thái đã cập nhật với tin nhắn mới được thêm vào danh sách tin nhắn.
 */
async function generate(
  state: typeof GraphState.State
): Promise<Partial<typeof GraphState.State>> {
  console.log("---GENERATE---");

  const { messages } = state;
  const question = messages[0].content as string;
  // Extract the most recent ToolMessage
  const lastToolMessage = messages
    .slice()
    .reverse()
    .find((msg) => msg._getType() === "tool");
  if (!lastToolMessage) {
    throw new Error("No tool message found in the conversation history");
  }

  const docs = lastToolMessage.content as string;

  const prompt = await pull<ChatPromptTemplate>("rlm/rag-prompt");

  const llm = new ChatOpenAI({
    model: "gpt-4o",
    temperature: 0,
    streaming: true,
  });

  const ragChain = prompt.pipe(llm);

  const response = await ragChain.invoke({
    context: docs,
    question,
  });

  return {
    messages: [response],
  };
}

async function main() {
  // Load environment variables
  dotenv.config({ path: "../shared/.env" });

  const OPENAI_API_KEY = process.env.OPENAI_API_KEY;
  if (!OPENAI_API_KEY) {
    throw new Error("OPENAI_API_KEY not found in environment variables");
  }

  const PdfPath = "../documents/2404.16130v1_Graph-RAG.pdf";
  const retriever = await loadAndProcessDocuments(PdfPath);

  const tool = createRetrieverTool(retriever, {
    name: "retrieve_graph_rag",
    description: "Search and return information about Graph-RAG.",
  });
  const tools = [tool];

  const toolNode = new ToolNode<typeof GraphState.State>(tools);

  // Define the graph
  const workflow = new StateGraph(GraphState)
    .addNode("agent", async (state) => agent(state, tools))
    .addNode("retrieve", toolNode)
    .addNode("gradeDocuments", gradeDocuments)
    .addNode("rewrite", rewrite)
    .addNode("generate", generate);

  // Call agent node to decide to retrieve or not
  workflow.addEdge(START, "agent");

  // Decide whether to retrieve
  workflow.addConditionalEdges(
    "agent",
    // Assess agent decision
    shouldRetrieve
  );

  workflow.addEdge("retrieve", "gradeDocuments");

  // Edges taken after the `action` node is called.
  workflow.addConditionalEdges(
    "gradeDocuments",
    // Assess agent decision
    checkRelevance,
    {
      // Call tool node
      yes: "generate",
      no: "rewrite", // placeholder
    }
  );

  workflow.addEdge("generate", END);
  workflow.addEdge("rewrite", "agent");

  // Compile
  const app = workflow.compile();

  const inputs = {
    messages: [new HumanMessage("What is Graph-RAG?")],
  };
  let finalState;
  for await (const output of await app.stream(inputs)) {
    for (const [key, value] of Object.entries(output)) {
      const lastMsg = output[key].messages[output[key].messages.length - 1];
      console.log(`Output from node: '${key}'`);
      console.dir(
        {
          type: lastMsg._getType(),
          content: lastMsg.content,
          tool_calls: lastMsg.tool_calls,
        },
        { depth: null }
      );
      console.log("---\n");
      finalState = value;
    }
  }

  console.log(JSON.stringify(finalState, null, 2));
}

main().catch(console.error);
